Adding documentation about bridging proposals to destination chains. (#1)
ScopeLift/uniswap-docs-fork: docs/concepts/multi-chain-proposals.md

> ---

Reviewer: [arbitrary comment location]

multi-chain proposals aren't showing up in the docs search (from http://localhost:3000/) -- do we need to manually index it, or would it be automatically picked up after deploy?

Author: I think this won't happen till after deploy

id: multi-chain-proposals

> title: Multi-Chain Proposals

Reviewer: [arbitrary comment location]

currently "multichain proposals" is listed on the nav after "Glossary." But it should be in "Governance," right?

<img width="273" alt="Screenshot 2025-01-10 at 10 30 08 AM" src="https://github.com/user-attachments/assets/315c632e-dbb1-40db-87e4-f172f7f447e8" />

Author: Done: c92c9da1c18c344ee66c52113ef08d3d591d08d2

---

> This is a living document which represents the current process guidelines for developing and advancing multi-chain Uniswap Governance Proposals. It was last updated Novenber 2024.

Reviewer: todo: change date

Author: Done: [4f93dbb2b2b43ae4f654f4f6e0c998c422a9a380](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/dec65b2cff02119fa56bcd1e2c49e92cc7681732)

> ## Introduction

Reviewer: can we combine Introduction into Overview? I think there's also opportunity to shorten these sections, we can take another pass together after other feedback is addressed.

Author: Done, basically combine them.. interested in your thoughts on how to shorten them: [2433e7a](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/2433e7adaa9075b0448213743290608738e43447)

This document provides a detailed guide on constructing such proposals for each target non-mainnet chain, including Solidity code snippets for each general approach.
The code snippets are written as Forge/Foundry scripts, and assume they are being run on the Forge platform, in a code repository with the necessary libraries (forge-std, uniswap-v3-core, and uniswap-v3-periphery) imported.
It is important to note that the code provided here is a general template and may require modification to suit the specific requirements of the target chain.

## Overview

Of the 24 chains where Uniswap V3 is deployed there are 10 chains that are OP Stack based and use an `L1CrossDomainMessenger` contract for bridging proposals.
There are 6 chains that use Wormhole for bridging proposals.
The remaining 8 chains use a variety of other methods for bridging proposals, all described below.

The table below contains links to the various proposal methods for each chain.

> | Proposal Method                                       | Chain(s)                                                                             |

Reviewer: i think this works for me, but i was thinking of a table like:

| Chain | Proposal Method |
| --- | --- |
| Arbitrum | ... |
| Base | ... |
| Blast | ... |
| BNB Chain | ... |
| Boba | ... |
| ... | ... |

in other words, index by the key that the user has (the chain they want to propose on), rather than the info we've provided

open to discussion here, lmk your thoughts!

Author: This approach is much better

Author: Done: [0fa67d0](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/0fa67d04c20bbc36d4cfaeb7536a6f3b5b127948)

| [Filecoin - TODO](#filecoin)                          | Filecoin                                                                             |
| [Op Stack - CrossChainAccount](#op-stack)             | Base, Blast, Boba, Linea, Manta Pacific, Mantle, Optimism, Redstone, WorldCoin, Zora |
| [Polygon - FxChild](#polygon)                         | Polygon                                                                              |
| [Polygon zkEVM - TODO](#polygon-zkevm)                | Polygon zkEVM                                                                        |
| [Scroll - CrossChainAccount](#scroll)                 | Scroll                                                                               |
| [Taiko - InvokableAccount](#taiko)                    | Taiko                                                                                |
| [Wormhole](#wormhole)                                 | BNB Chain, Celo, Gnosis Chain, Moonbeam, Rootstock, Sei                              |
| [ZkSync Era - Aliased Timelock](#zksync-era)          | ZkSync Era                                                                           |

## Arbitrum

> Arbitrum uses an approach where the owner of the V3Factory is a special aliased address (offset by the value 0x1111000000000000000000000000000000001111) that (when the offset is subtracted away) is the L1 address of the Uniswap DAO Timelock contract. A Solidity code example for sending a proposal to Arbitrum is shown below.

Reviewer: here and elsewhere, can we code-format values and contract names?

```suggestion
Arbitrum uses an approach where the owner of the V3Factory is a special aliased address (offset by the value `0x1111000000000000000000000000000000001111`) that (when the offset is subtracted away) is the L1 address of the Uniswap DAO Timelock contract. A Solidity code example for sending a proposal to Arbitrum is shown below.
```

Author: Done: [a85dd16dd0fc6df67865fb15d89fda3b5c49d072](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/dc1c1caee8f5b5c6b92d0e33641419a8244230a7)

Proposal calldata on mainnet intended for Arbitrum should be forwarded through a call to `createRetryableTicket` on the Arbitrum Inbox contract.
The call would have wrapped calldata for a Uniswap contract function call that would effect the change. An example:

> ```

Reviewer: did you explore linking out to these instead?

Author: I did this like other parts of the docs repo.. put the code in the doc, but also provided github links to the source files themselves: b96e51a41d6553207a2462e9958acda5bd2857ab

            callValueRefundAddress,
            maxGas,
            gasPriceBid,
            _v3FactoryEnableFeeAmounCalldata
        );

        // make the proposal on the L1 side
        IUniswapGovernorBravoDelegator governor = IUniswapGovernorBravoDelegator(uniswapGovernorBravoDelegatorAddress);
        address[] memory _targets = new address[](1);
        uint256[] memory _values = new uint256[](1);
        string[] memory _signatures = new string[](1);
        bytes[] memory _calldatas = new bytes[](1);
        _targets[0] = INBOX_ADDRESS;
        _values[0] = 0;
        _signatures[0] = "createRetryableTicket(address,uint256,uint256,address,address,uint256,uint256,bytes)";
        _calldatas[0] = _proposalCalldata;
        uint256 _proposalId = governor.propose(
            _targets,
            _values,
            _signatures,
            _calldatas,
            "Proposal to enable 10000 fee amount of 205 on Uniswap V3 Factory on Arbitrum"
        );
        console2.log("Proposal ID: %d", _proposalId);
    }
}
```

## Avalanche

Avalanche makes use of a contract called `OmnichainGovernanceExecutor` to receive proposals from Ethereum mainnet.

> Proposal calldata on mainnet intended for Avalanche should be forwarded through a call to `send` on the contract called LayerZero:EndpointV2.

Reviewer: example of a contract name that could be code-formatted.

```suggestion
Proposal calldata on mainnet intended for Avalanche should be forwarded through a call to `send` on the contract called `LayerZero:EndpointV2`.
```

up for discussion, lmk!

Author: Done: [a85dd16dd0fc6df67865fb15d89fda3b5c49d072](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/dc1c1caee8f5b5c6b92d0e33641419a8244230a7)

            0 // Parameters for adapter services (e.g., gas limits)
        );

        // make the proposal on the L1 side
        IUniswapGovernorBravoDelegator governor = IUniswapGovernorBravoDelegator(uniswapGovernorBravoDelegatorAddress);
        address[] memory _targets = new address[](1);
        uint256[] memory _values = new uint256[](1);
        string[] memory _signatures = new string[](1);
        bytes[] memory _calldatas = new bytes[](1);
        _targets[0] = LAYERZERO_ENDPOINT_ADDRESS;
        _values[0] = 0;
        _signatures[0] = "send(uint16,bytes,bytes,address,address,bytes)";
        _calldatas[0] = _sendCalldata;
        uint256 _proposalId = governor.propose(
            _targets,
            _values,
            _signatures,
            _calldatas,
            "Proposal to enable 10000 fee amount of 205 on Uniswap V3 Factory on AVAX"
        );
        console2.log("Proposal ID: %d", _proposalId);
    }
}
```

## Filecoin

Filecoin EVM ?? TODO: Add details

## OP Stack

Proposal calldata on mainnet meant for an OP stack chain should be forwarded through a call to sendMessage on the chain's corresponding mainnet L1CrossDomainMessenger contract.

> The call would have doubly-wrapped calldata for the CrossChainAccount `forward` function as well as the ultimate Uniswap contract function call that would effect the change. An example for Optimism:

Reviewer: ```suggestion
The call would have doubly-wrapped calldata for the `CrossChainAccount:forward` function as well as the ultimate Uniswap contract function call that would effect the change. An example for Optimism:
```

Author: Done: [a85dd16dd0fc6df67865fb15d89fda3b5c49d072](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/dc1c1caee8f5b5c6b92d0e33641419a8244230a7)

        // encode the calldata for the CrossChainAccount.forward call
        bytes memory _messageForwardCalldata = abi.encodeWithSelector(
            ICrossChainAccount.forward.selector, v3FactoryTargetAddress, _v3FactoryEnableFeeAmounCalldata
        );

        // make the proposal on the L1 side
        IUniswapGovernorBravoDelegator governor = IUniswapGovernorBravoDelegator(uniswapGovernorBravoDelegatorAddress);
        address[] memory _targets = new address[](1);
        uint256[] memory _values = new uint256[](1);
        string[] memory _signatures = new string[](1);
        bytes[] memory _calldatas = new bytes[](1);
        _targets[0] = l1CrossDomainMessengerAddress;
        _values[0] = 0;
        _signatures[0] = "sendMessage(address,bytes,uint256)";
        _calldatas[0] = abi.encode(address(crossChainAccountTargetAddress), _messageForwardCalldata, 500_000);
        uint256 _proposalId = governor.propose(
            _targets,
            _values,
            _signatures,
            _calldatas,
            "Proposal to enable 10000 fee amount of 205 on Uniswap V3 Factory on Optimism"
        );
        console2.log("Proposal ID: %d", _proposalId);
    }
}
```

> Because 10 networks all use the same CrossChainAccount approach for communication, the only modifications to the above code snippet for those networks would be to change the first 3 address constant definitions. The definitions for all of the OP Stack based networks are below:

Reviewer: ```suggestion
Because 10 networks all use the same CrossChainAccount approach for communication, the only modifications to the above code snippet for those networks would be to change the first 3 address constant definitions.

:::warning

**Linea**'s `sendMessage` reverses the order of the `fee` and `message` parameters.

The relevant OP Stack contract addresses are as follows:
```

Author: Done: [769757310e8919e37e364cdbf7f703ef38ffa7c8](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/028fb68f179fec96f2a3296fc54893e43ce0ce60)

> | Network       | l1CrossDomainMessengerAddress              | v3FactoryTargetAddress                     | crossChainAccountTargetAddress             |

Reviewer: nit here: this table is wide on my browser. what about each row having only one cell with all 3 contract addresses?

e.g. 

| Network | Contract addresses |
| --- | --- |
| Base |  `l1CrossDomainMessenger`: &nbsp;`0x866E82a600A1414e583f7F13623F1aC5d58b0Afa`<br/>`v3FactoryTarget`: &nbsp;`0x33128a8fC17869897dcE68Ed026d694621f6FDfD`<br/>`crossChainAccountTarget`: &nbsp;`0x31FAfd4889FA1269F7a13A66eE0fB458f27D72A9` |
| Blast | ... |

meh, that doesn't look good either...

what about getting rid of the `Address` suffix on our table headings, and code-formatting each address? maybe that'll get us some of the way there.

Author: Done (went with an html table): [769757310e8919e37e364cdbf7f703ef38ffa7c8](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/028fb68f179fec96f2a3296fc54893e43ce0ce60)

| ------------- | ------------------------------------------ | ------------------------------------------ | ------------------------------------------ |
| Base          | 0x866E82a600A1414e583f7F13623F1aC5d58b0Afa | 0x33128a8fC17869897dcE68Ed026d694621f6FDfD | 0x31FAfd4889FA1269F7a13A66eE0fB458f27D72A9 |
| Blast         | 0x5D4472f31Bd9385709ec61305AFc749F0fA8e9d0 | 0x792edAdE80af5fC680d96a2eD80A44247D2Cf6Fd | 0x2339C0d23b60739B3E5ABF201F05903D24A26C77 |
| Boba          | 0x6D4528d192dB72E282265D6092F4B872f9Dff69e | 0xFFCd7Aed9C627E82A765c3247d562239507f6f1B | 0x53163235746CeB81Da32293bb0932e1A599256B4 |

> | Linea - 1     | 0xd19d4B5d358258f05D7B411E21A1460D11B0876F | 0x31FAfd4889FA1269F7a13A66eE0fB458f27D72A9 | 0x581F86Da293A1D5Cd087a10E7227a75d2d2201A8 |

Reviewer: ```suggestion
| Linea (see warning)     | 0xd19d4B5d358258f05D7B411E21A1460D11B0876F | 0x31FAfd4889FA1269F7a13A66eE0fB458f27D72A9 | 0x581F86Da293A1D5Cd087a10E7227a75d2d2201A8 |
```

Author: Done: [769757310e8919e37e364cdbf7f703ef38ffa7c8](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/028fb68f179fec96f2a3296fc54893e43ce0ce60)

| Manta Pacific | 0x635ba609680c55C3bDd0B3627b4c5dB21b13c310 | 0x06D830e15081f65923674268121FF57Cc54e4e23 | 0x683553d74D9779955a15d57D208234C956B6Eae6 |
| Mantle        | 0x676A795fe6E43C17c668de16730c3F690FEB7120 | 0x0d922Fb1Bc191F64970ac40376643808b4B74Df9 | 0x9b7aC6735b23578E81260acD34E3668D0cc6000A |
| Optimism      | 0x25ace71c97B33Cc4729CF772ae268934F7ab5fA1 | 0x1F98431c8aD98523631AE4a59f267346ea31F984 | 0xa1dD330d602c32622AA270Ea73d078B803Cb3518 |
| Redstone      | 0x592C1299e0F8331D81A28C0FC7352Da24eDB444a | 0xece75613Aa9b1680f0421E5B2eF376DF68aa83Bb | 0x2d00e94d78Fc307FC5E6195BBe2fB6aFC2FC07d4 |
| Worldcoin     | 0xf931a81D18B1766d15695ffc7c1920a62b7e710a | 0x7a5028BDa40e7B173C278C5342087826455ea25a | 0xcb2436774C3e191c85056d248EF4260ce5f27A9D |
| Zora          | 0xdC40a14d9abd6F410226f1E6de71aE03441ca506 | 0x7145F8aeef1f6510E92164038E1B6F8cB2c42Cbb | 0x36eEC182D0B24Df3DC23115D64DB521A93D5154f |

> 1 - Slightly different sending code for Linea, with the 'sendMessage' function's `fee` and `message` parameters swapped.

Reviewer: ```suggestion
```

Author: Done: [769757310e8919e37e364cdbf7f703ef38ffa7c8](https://github.com/ScopeLift/uniswap-docs-fork/pull/1/commits/028fb68f179fec96f2a3296fc54893e43ce0ce60)

In addition to the proposal message parameters, the `sendMessage` function of this contract also takes both a target address for the message receiving contract on the non-mainnet chain, as well as its Chain ID as parameters, allowing the function to be used for sending to multiple destination chains. 
A Solidity code example for sending a proposal to one these chains (BNB Chain) is provided [below](#wormhole-proposal-bridging-example).
This code can be used for sending proposals to the other 3 chains (Gnosis Chain, Moonbeam, and Rootstock), with the only changes needed being the actual depoloyed address constants for the targeted contracts.

Each of the sections below provides a high-level description of the method used to bridge proposals to each target non-mainnet chain.

#### Arbitrum Cross-chain Communication

Arbitrum uses an approach where the owner of the V3Factory is a special aliased address (offset by the value 0x1111000000000000000000000000000000001111) that (when the offset is subtracted away) is the L1 address of the Uniswap DAO Timelock contract. A Solidity code example for sending a proposal to Arbitrum is shown [below](#arbitrum-proposal-bridging-details).

> #### Avalanch/AVAX Cross-chain Communication

Reviewer: ```suggestion
#### Avalanche Cross-chain Communication
```
